File: python/stewardship/force_resolve.py
#! /usr/bin/env python3

import json
import os
import sys
from pathlib import Path

from senzing import SzEngineFlags, SzError
from senzing_core import SzAbstractFactory

ENGINE_CONFIG_JSON = os.getenv("SENZING_ENGINE_CONFIGURATION_JSON", "{}")
INSTANCE_NAME = Path(__file__).stem

PURGE_MSG = """
**************************************** WARNING ****************************************

This example will purge all currently loaded data from the Senzing datastore!
Before proceeding, all instances of Senzing (custom code, tools, etc.) must be shut down.

*****************************************************************************************

Are you sure you want to continue and purge the Senzing datastore? Type YESPURGESENZING to purge: """

RECORDS = [
    {
        "DATA_SOURCE": "TEST",
        "RECORD_ID": "1",
        "PRIMARY_NAME_FULL": "Patrick Smith",
        "AKA_NAME_FULL": "Paddy Smith",
        "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
        "PHONE_NUMBER": "787-767-2688",
        "DATE_OF_BIRTH": "1/12/1990",
    },
    {
        "DATA_SOURCE": "TEST",
        "RECORD_ID": "2",
        "PRIMARY_NAME_FULL": "Patricia Smith",
        "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
        "PHONE_NUMBER": "787-767-2688",
        "DATE_OF_BIRTH": "5/4/1994",
    },
    {
        "DATA_SOURCE": "TEST",
        "RECORD_ID": "3",
        "PRIMARY_NAME_FULL": "Pat Smith",
        "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
        "PHONE_NUMBER": "787-767-2688",
    },
]

if input(PURGE_MSG) != "YESPURGESENZING":
    sys.exit()

try:
    sz_factory = SzAbstractFactory(INSTANCE_NAME, ENGINE_CONFIG_JSON, verbose_logging=False)
    sz_diagnostic = sz_factory.create_diagnostic()
    sz_engine = sz_factory.create_engine()
    sz_diagnostic.purge_repository()

    print()
    for record in RECORDS:
        data_source = record["DATA_SOURCE"]
        record_id = record["RECORD_ID"]
        sz_engine.add_record(data_source, record_id, json.dumps(record))
        print(f"Record {record_id} added")

    print()
    for record_id in ("1", "2", "3"):
        response1 = sz_engine.get_entity_by_record_id(
            "TEST",
            record_id,
            SzEngineFlags.SZ_ENTITY_BRIEF_DEFAULT_FLAGS,
        )
        get_json = json.loads(response1)
        print(f"Record {record_id} currently resolves to entity" f" {get_json['RESOLVED_ENTITY']['ENTITY_ID']}")

    print("\nUpdating records with TRUSTED_ID to force resolve...\n")
    record1 = sz_engine.get_record("TEST", "1")
    record2 = sz_engine.get_record("TEST", "3")
    get1_json = json.loads(record1)
    get2_json = json.loads(record2)
    get1_json["JSON_DATA"].update({"TRUSTED_ID_NUMBER": "TEST_R1-TEST_R3", "TRUSTED_ID_TYPE": "FORCE_RESOLVE"})
    get2_json["JSON_DATA"].update({"TRUSTED_ID_NUMBER": "TEST_R1-TEST_R3", "TRUSTED_ID_TYPE": "FORCE_RESOLVE"})
    sz_engine.add_record("TEST", "1", json.dumps(get1_json["JSON_DATA"]))
    sz_engine.add_record("TEST", "3", json.dumps(get2_json["JSON_DATA"]))

    for record_id in ("1", "2", "3"):
        response2 = sz_engine.get_entity_by_record_id(
            "TEST",
            record_id,
            SzEngineFlags.SZ_ENTITY_BRIEF_DEFAULT_FLAGS,
        )
        get_json = json.loads(response2)
        print(f"Record {record_id} now resolves to entity" f" {get_json['RESOLVED_ENTITY']['ENTITY_ID']}")
except SzError as err:
    print(f"{err.__class__.__name__} - {err}", file=sys.stderr)

File: java/snippets/stewardship/ForceResolve.java
package stewardship;

import java.util.*;
import javax.json.*;
import java.io.*;

import com.senzing.sdk.*;
import com.senzing.sdk.core.SzCoreEnvironment;

import static com.senzing.sdk.SzFlag.*;

/**
 * Provides a simple example of force-resolving records that
 * otherwise will not resolve to one another.
 */
public class ForceResolve {
    private static final String TEST = "TEST";

    public static void main(String[] args) {
        // get the senzing repository settings
        String settings = System.getenv("SENZING_ENGINE_CONFIGURATION_JSON");
        if (settings == null) {
            System.err.println("Unable to get settings.");
            throw new IllegalArgumentException("Unable to get settings");
        }

        // create a descriptive instance name (can be anything)
        String instanceName = ForceResolve.class.getSimpleName();

        // initialize the Senzing environment
        SzEnvironment env = SzCoreEnvironment.newBuilder()
            .settings(settings)
            .instanceName(instanceName)
            .verboseLogging(false)
            .build();
        
        try {
            // get the engine from the environment
            SzEngine engine = env.getEngine();

            Map<SzRecordKey, String> recordMap = getRecords();
            // loop through the example records and add them to the repository
            for (Map.Entry<SzRecordKey,String> entry : recordMap.entrySet()) {
                SzRecordKey recordKey = entry.getKey();
                String recordDefinition = entry.getValue();
 
                // call the addRecord() function with no flags
                engine.addRecord(recordKey, recordDefinition, SZ_NO_FLAGS);

                System.out.println("Record " + recordKey.recordId() + " added");
                System.out.flush();
            }

            System.out.println();
            for (SzRecordKey recordKey : recordMap.keySet()) {
                String result = engine.getEntity(recordKey, SZ_ENTITY_BRIEF_DEFAULT_FLAGS);
                JsonObject jsonObj = Json.createReader(new StringReader(result)).readObject();
                long entityId = jsonObj.getJsonObject("RESOLVED_ENTITY")
                                       .getJsonNumber("ENTITY_ID").longValue();
                System.out.println(
                    "Record " + recordKey + " originally resolves to entity " + entityId);
            }
            System.out.println();
            System.out.println("Updating records with TRUSTED_ID_NUMBER to force resolve...");
            SzRecordKey key1 = SzRecordKey.of(TEST, "1");
            SzRecordKey key3 = SzRecordKey.of(TEST, "3");
            
            String record1 = engine.getRecord(key1, SZ_RECORD_DEFAULT_FLAGS);
            String record3 = engine.getRecord(key3, SZ_RECORD_DEFAULT_FLAGS);

            JsonObject obj1 = Json.createReader(new StringReader(record1)).readObject();
            JsonObject obj3 = Json.createReader(new StringReader(record3)).readObject();

            obj1 = obj1.getJsonObject("JSON_DATA");
            obj3 = obj3.getJsonObject("JSON_DATA");

            JsonObjectBuilder job1 = Json.createObjectBuilder(obj1);
            JsonObjectBuilder job3 = Json.createObjectBuilder(obj3);

            for (JsonObjectBuilder job : List.of(job1, job3)) {
                job.add("TRUSTED_ID_NUMBER", "TEST_R1-TEST_R3");
                job.add("TRUSTED_ID_TYPE", "FORCE_RESOLVE");
            }

            record1 = job1.build().toString();
            record3 = job3.build().toString();

            engine.addRecord(key1, record1, SZ_NO_FLAGS);
            engine.addRecord(key3, record3, SZ_NO_FLAGS);

            System.out.println();
            for (SzRecordKey recordKey : recordMap.keySet()) {
                String result = engine.getEntity(recordKey, SZ_ENTITY_BRIEF_DEFAULT_FLAGS);
                JsonObject jsonObj = Json.createReader(new StringReader(result)).readObject();
                long entityId = jsonObj.getJsonObject("RESOLVED_ENTITY")
                                       .getJsonNumber("ENTITY_ID").longValue();
                System.out.println(
                    "Record " + recordKey + " now resolves to entity " + entityId);
            }
            System.out.println();

        } catch (SzException e) {
            // handle any exception that may have occurred
            System.err.println("Senzing Error Message : " + e.getMessage());
            System.err.println("Senzing Error Code    : " + e.getErrorCode());
            e.printStackTrace();
            throw new RuntimeException(e);

        } catch (Exception e) {
            e.printStackTrace();
            if (e instanceof RuntimeException) {
                throw ((RuntimeException) e);
            }
            throw new RuntimeException(e);

        } finally {
            // IMPORTANT: make sure to destroy the environment
            env.destroy();
        }

    }

    /**
     * This is a support method for providing example records to add.
     * 
     * @return A {@link Map} of {@link SzRecordKey} keys to {@link String}
     *         JSON text values desribing the records to be added.
     */
    public static Map<SzRecordKey, String> getRecords() {
        Map<SzRecordKey, String> records = new LinkedHashMap<>();
        records.put(
            SzRecordKey.of("TEST", "1"),
            """
            {
                "DATA_SOURCE": "TEST",
                "RECORD_ID": "1",
                "PRIMARY_NAME_FULL": "Patrick Smith",
                "AKA_NAME_FULL": "Paddy Smith",
                "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
                "PHONE_NUMBER": "787-767-2688",
                "DATE_OF_BIRTH": "1/12/1990"
            }
            """);
        
        records.put(
            SzRecordKey.of("TEST", "2"),
            """
            {
                "DATA_SOURCE": "TEST",
                "RECORD_ID": "2",
                "PRIMARY_NAME_FULL": "Patricia Smith",
                "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
                "PHONE_NUMBER": "787-767-2688",
                "DATE_OF_BIRTH": "5/4/1994"
            }
            """);
        
        records.put(
            SzRecordKey.of("TEST", "3"),
            """
            {
                "DATA_SOURCE": "TEST",
                "RECORD_ID": "3",
                "PRIMARY_NAME_FULL": "Pat Smith",
                "ADDR_FULL": "787 Rotary Dr, Rotorville, RI, 78720",
                "PHONE_NUMBER": "787-767-2688"
            }
            """);
       
        return records;
    }
}